data/heuristics.go

```diff
@@ -0,0 +1,97 @@
+package data
+
+import "regexp"
+
+type (
+	Heuristics []Matcher
+
+	Matcher interface {
+		Match(data []byte) bool
+	}
+
+	Languages struct {
+		langs []string
+	}
+
+	Rule interface {
+		GetLanguages() []string
```

Comment:
Separately, it's not clear why a `Rule` has `Languages` and nothing else. I mean of course I understand why it is this way operationally, but the semantics are kind of disjoint: We're saying something about the concrete type we know inhabits this interface, rather than the meaning this package cares about.

This may wind up being a larger design question, in which case we can defer it.

Reply:
> Separately, it's not clear why a Rule has Languages and nothing else

I'm totally open to suggestions on a better names for these. 
But every rule, if it matches, can produce a number of languages thus the names in here.

Reply:
I think the source of confusion here is twofold:
 - removing `Get` prefix from `GetLanguages` will not allow the compiler to distinguish between `Languages` type and `Languages()` method. This will result in a panic at a runtime.
 - As Go does not allow all Rules to inherit a data member `langs []string`, it's a trick that I use here: type `Languages` does implement abstract interface `Rule`, and then every concrete `*Rule` embeds the `Languages` type and thus:
     * implements the Rule interface
     * gets access to a data member though `GetLanguages()`

Please let me know if there is a better way of modeling this, but so far this one seems the most concise to me given the language constraints.

I belive this all might also be resolved just by picking a better names for each of those 3 entities mentioned above.

Reply:
There shouldn't be any conflict between the method name and a type name, so even if it were simply `type Rule interface { Languages() []string }` that shouldn't cause problems I think.

That said: My main confusion I think is about names rather than basic structure. If we do not need this interface except to type-assert variant implementations, one simple thing we can do is unexport the interface so it's not part of the public API for the package. In fact, if it's only used in that one place, you can even drop the name:

```go
langs := matcher.(interface{ Languages() []string }).Languages()

// or alternatively
langs := matcher.(interface {
	Languages() []string
}).Languages()
```

…if you want. Though some folks find that harder to read, so YMMV.

Unexporting `rule` would probably assuage most of my concern, which is about clutter in the public API.

Reply:
Thank you for suggestions and explicitly sharing main concern!

I belive it has been addressed in 97ab29a